c8d: Implement push

vvoland opened the pull request:
- upstreams: https://github.com/rumpl/moby/pull/2
- upstreams: https://github.com/rumpl/moby/pull/74
- upstreams: https://github.com/rumpl/moby/pull/80
- upstreams: https://github.com/rumpl/moby/pull/15


Implements `docker push` under containerd image store.  When
pushing manifest lists that reference a content which is not present in
the local content store, it will attempt to perform the cross-repo mount
the content if possible.

Considering this scenario:

```bash
$ docker pull docker.io/library/busybox
```
This will download manifest list and only host platform-specific
manifest and blobs.

Note, tagging to a different repository (but still the same registry) and pushing:
```bash
$ docker tag docker.io/library/busybox docker.io/private-repo/mybusybox
$ docker push docker.io/private-repo/mybusybox
```

will result in error, because the neither we nor the target repository
doesn't have the manifests that the busybox manifest list references
(because manifests can't be cross-repo mounted).

If for some reason the manifests and configs for all other platforms
would be present in the content store, but only layer blobs were
missing, then the push would work, because the blobs can be cross-repo
mounted (only if we push to the same registry).


The `container.io/distribution.source` label is taken only from the target blob we are pushing. It's a change from fork's behavior.
In fork for each missing content we did search the entire content store for a manifest that includes it and took it's distribution source label:
https://github.com/moby/moby/blob/2670c94d8298523ca90db1b097d0c4ebfc3ccc56/daemon/containerd/lazy_push.go#L367-L489

This was mostly a workaround for buildkit image exporter not fetching all platform-specific content (only the host's platform). Eg. doing `FROM 'ubuntu' | docker buildx build --platform <host>,<other> -` would only fetch ubuntu rootfs for `<host>` platform, but not for `<other>`. Because the built image didn't have any `distribution.source` that would point to the original `docker.io/library/ubuntu` - it was impossible to know where to obtain the blob from, based on the built image alone. This was only possible by traversing all manifests in content store, checking if they have a `distribution.source` label and then traversing them to check if they happen to define this blob.
Since this is no longer needed as buildkit already does `store=true` by default, I decided to change this.

**- What I did**
Implemented push

**- How I did it**
Use containerd remote to push the content
Some missing content can be cross-repo mounted. In this case fake their existence by wrapping the `content.Store` to return fake `content.Info` that includes the `containerd.io/distribution.source` label with the source repository.
That makes the push not end with NotFound error and allows it to perform cross-repo mount.

<details>

<summary>**How to verify it [click to expand]**</summary>

TODO: New examples
</details>


**- A picture of a cute animal (not mandatory but encouraged)**
![mobbunny](https://user-images.githubusercontent.com/5046555/217814826-56c699b1-ff48-4858-8e4c-7a11fc74f783.png)

vvoland committed:
c8d: Implement push

This implements `docker push` under containerd image store.  When
pushing manifest lists that reference a content which is not present in
the local content store, it will attempt to perform the cross-repo mount
the content if possible.

Considering this scenario:

```bash
$ docker pull docker.io/library/busybox
```
This will download manifest list and only host platform-specific
manifest and blobs.

Note, tagging to a different repository (but still the same registry) and pushing:
```bash
$ docker tag docker.io/library/busybox docker.io/private-repo/mybusybox
$ docker push docker.io/private-repo/mybusybox
```

will result in error, because the neither we nor the target repository
doesn't have the manifests that the busybox manifest list references
(because manifests can't be cross-repo mounted).

If for some reason the manifests and configs for all other platforms
would be present in the content store, but only layer blobs were
missing, then the push would work, because the blobs can be cross-repo
mounted (only if we push to the same registry).

Signed-off-by: Paweł Gronowski <pawel.gronowski@docker.com>

vvoland committed:
images/push: Accept reference

Push the reference parsing from repo and tag names into the api and pass
a reference object to the ImageService.

Signed-off-by: Paweł Gronowski <pawel.gronowski@docker.com>

vvoland commented on daemon/containerd/image_push.go:
```diff
@@ -2,14 +2,247 @@ package containerd
+		return nil, err
+	}
+
+	// Set GC labels recursively for all downloaded blobs.
+	setLabels := containerdimages.SetChildrenLabels(store, containerdimages.ChildrenHandler(store))
+	err = containerdimages.Dispatch(ctx, setLabels, nil, target)
+	if err != nil {
+		logrus.WithError(err).Warn("set children gc labels failed")
+	}
```
Not sure about this one - it persist downloaded content if it couldn't be cross repo mounted.
In case of pushing shallowly pulled image to a different registry, this will effectively unshallow an image and make it possible to use other downloaded platform.
On the other hand, deleting them just after we finish push is wasteful.

This should probably be an opt-in (once we have a CLI flag for that).

vvoland commented:
Arghh, wanted to make the PushImage accept `reference.NamedTagged`, but forgot about `--all-tags` case in which there's no tag at all 🤦🏻  
```
=== Failed
=== FAIL: amd64.integration-cli TestDockerRegistrySuite/TestPushMultipleTags (0.16s)
    docker_cli_push_test.go:79: assertion failed: 
        Command:  /usr/local/cli/docker push 127.0.0.1:5000/dockercli/busybox
        ExitCode: 1
        Error:    exit status 1
        Stdout:   The push refers to repository [127.0.0.1:5000/dockercli/busybox]
        
        Stderr:   tag does not exist: 127.0.0.1:5000/dockercli/busybox:latest
        
        
        Failures:
        ExitCode was 1 expected 0
        Expected no error
    check_test.go:384: [deae1ffcfb672] daemon is not started
    --- FAIL: TestDockerRegistrySuite/TestPushMultipleTags (0.16s)
```

AkihiroSuda commented:
I'm not sure whether it is safe to push a manifest without its layer blobs. 
From my comment in nerdctl, some registry doesn't/didn't accept it?

https://github.com/containerd/nerdctl/blob/b0701ba875880db5a6e7585df148f0eaea36283a/pkg/cmd/image/push.go#L92-L106

> // Push fails with "400 Bad Request" when the manifest is multi-platform but we do not locally have multi-platform blobs.
> // So we create a tmp reduced-platform image to avoid the error.

AkihiroSuda commented on daemon/containerd/image_push.go:
```diff
@@ -2,14 +2,251 @@ package containerd
+	return *s, nil
+}
+
+const labelDistributionSource = "containerd.io/distribution.source."
```
nit: The const should be publicized on containerd side: https://github.com/containerd/containerd/blob/97480afdac09c947d48f5e3a134db86c78f4bfa6/remotes/docker/handler.go#L35

Can be done later though

vvoland replied:
https://github.com/containerd/containerd/pull/8224

tianon replied:
Does containerd append these labels on push, too? (or only pull?)

If I build something, push it somewhere, then build something else based on that first thing and push it to the same registry, it can use cross-repo blob mounts too. :sweat_smile:

(see also the related https://github.com/moby/moby/pull/44757 :joy: :see_no_evil:)

vvoland replied:
Nope, containerd only appends these on pull.
But I just made this code do it and it seems to work!

quick test:
```bash
$ docker run --name test -it alpine
# cat /dev/random >/file  # ctr-c after a while
$ docker commit test 172.172.0.2:5000/big-image
$ docker push 172.172.0.2:5000/big-image
$ docker run --name test2 -it 172.172.0.2:5000/big-image
# apk add fish
...
$ docker commit test2 172.172.0.2:5000/big-image-with-small-change
$ docker push 172.172.0.2:5000/big-image-with-small-change
Using default tag: latest
f8d2deb24c5c: Pushed
c41833b44d91: Pushed
8b626a5feef1: Pushed
140adfd2fa47: Pushed
c76cc252a7ff: Pushed # <- pushes only this changed layer
```

The status is still `Pushed` though because the containerd doesn't actually expose that information (I'm working on a containerd PR for that one though).

tianon replied:
That's really neat!!  Thank you for testing/working on it :bow:

(For interested parties, I think https://github.com/containerd/containerd/pull/8330 is the relevant containerd PR :eyes:)

vvoland commented:
> I'm not sure whether it is safe to push a manifest without its layer blobs.

Yes, registry (DockerHub at least) will reject the manifest because it references blobs not present in the repository.
The reason why I do `prepareMissing` only for manifest lists and not manifests, is that I don't expect that image which targets platform-specific manifest will have any blobs missing.

Removing that check for manifests wouldn't hurt though, but if there's a case in which the one-platform image has layers missing then probably there's something wrong going on.

thaJeztah commented on api/server/router/image/image_routes.go:
```diff
@@ -153,7 +153,25 @@ func (ir *imageRouter) postImagesPush(ctx context.Context, w http.ResponseWriter
+
+	var ref reference.Named
+
+	// Tag is empty only in case ImagePushOptions.All is true.
+	if tag != "" {
```
I just recalled we still have an open ticket for this to change this at the API level as well;

- https://github.com/moby/moby/pull/40302 changed the client to require an explicit opt-in for "all tags"
- https://github.com/docker/cli/pull/2220 changed the CLI to use it
- https://github.com/moby/moby/issues/42053has a proposal to change the API to have an explicit `all-tags` options

Maybe we should start working on making that API change 🤔 (we could mention the linked ticket here as a TODO)

vvoland commented on daemon/containerd/image_push.go:
```diff
@@ -2,14 +2,251 @@ package containerd
-	return errdefs.NotImplemented(errors.New("not implemented"))
+// PushImage initiates a push operation of the image pointed to by targetRef.
+// If not all content referenced by the image manifest list is available
+// locally, it will attempt to cross-repo mount them or download if the former
+// is not possible.
```
No. Once we have CLI flags, this should definitely be opt-in.

rumpl commented on daemon/containerd/image_push.go:
```diff
@@ -2,14 +2,247 @@ package containerd
+	progress *pushProgress, limiter *semaphore.Weighted,
+) (map[digest.Digest]distributionSource, error) {
+	mountableBlobs := map[digest.Digest]distributionSource{}
+	source, err := getDigestSource(ctx, store, target.Digest)
```
What if a user explicitly sets `store=false`?

tianon commented on daemon/containerd/image_push.go:
```diff
@@ -2,15 +2,277 @@ package containerd
+	if containerdimages.IsIndexType(mediaType) {
+		return false
+	}
+	if containerdimages.IsConfigType(mediaType) {
+		return false
+	}
```
This one is a little over-zealous -- manifest objects _are_ special (indexes and image manifests), but config objects are just regular blobs (like layers), so if they already exist, they're 100% fair game for cross-blob mounts :eyes:

vvoland replied:
Ah, indeed. I was sure that configs are uploaded as manifests too, but this was probably some mistake when developing the initial fork version. Thanks!

tianon commented on api/swagger.yaml:
```diff
@@ -8490,6 +8490,19 @@ paths:
           in: "query"
           description: "The tag to associate with the image on the registry."
           type: "string"
+        - name: "platform"
+          in: "query"
+          description: |
+            Platform in the format `os[/arch[/variant]]` used to select
+            a platform-specific manifest if the image is an index/manifest list.
+
+            If the option is not set, then no specific manifest is select and the
+            whole image index is pushed.
+
+            When specified, the daemon checks if the image has any manifest that
+            matches the requested platform and pushes it instead of the whole index.
+            If there's no matching manifest it returns a `404` status.
+          type: "string"
```
As I've noted in my other comment, IMO we should spend a bit more time thinking about the UX we want from `docker push` WRT platforms before we implement this in the API ("no is temporary, yes is forever"), but I seem to recall being in a minority on that in the maintainers call where we discussed this, so I'm not going personally to block this if other maintainers don't agree. :bow:

tianon reviewed (commented):
Overall this looks pretty good to me -- in the maintainers call where we discussed this, I was concerned about having `docker push` end up _pulling_ content locally, but just doing as much cross-blob mounting as we can for objects we don't have locally seems 100% sane and honestly really, really clever :+1:

I've got a few of what I think are minor nits and would love to hear more maintainer opinions about exposing `platform` in the API for this endpoint before we've really ironed out the UX (because it's much harder to walk that external API back later if we decide we want something different than it is to just not introduce it for now), but I don't feel so strongly as to block it if I'm alone in my concerns. :bow: :heart:

tianon commented:
Discussed in the maintainers meeting today, and we'd like to hold off on changes to the push API (specifically the addition of `platform`) until we have the larger UX discussion about how push should look in a world where we have a more complex underlying image store and how we can better take advantage of that. :bow:

I think concretely that means pulling e39998f70c2ecd2d369ae7da29ff78f11937f963 out for now (and the related code in the previous commit). :eyes:

vvoland commented on daemon/images/image_push.go:
```diff
@@ -14,20 +14,8 @@ import (
-		return err
-	}
-	if tag != "" {
-		// Push by digest is not supported, so only tags are supported.
```
It probably requires an API change too though? I'm guessing we don't want the `tag` to also be able to specify digests? 👀 
Possibly, we could think about deprecating separate `repo` and `tag` arguments (but still support it), and have `ref` instead as a preferred way to reference an image?

vvoland commented on daemon/containerd/image_push.go:
```diff
@@ -2,15 +2,277 @@ package containerd
+	}))
+	defer finishProgress()
+
+	var limiter *semaphore.Weighted = nil // TODO: Respect max concurrent downloads/uploads
```
Actually we can support it for push just fine because we use the `remotes.PushContent` directly. It's the pull that uses the c8d's `client.Push` method that can't use the same limiter for multiple concurrent operations. 
For pull, I expect that at some point we would have the need to do something more than the c8d `client.Pull` provides and we will have to use the lower level functions anyway, which would allow us to support MaxConcurrentDownloads too.
So perhaps it would be fine (at least for now) to only have "MaxConcurrentUploads" implemented?

tianon replied:
Yeah, that seems reasonable to me, although IMO doesn't have to be part of _this_ PR unless you think the implementation is really straightforward/quick. :bow:

Given this isn't blocked on upstream (as I'd mistakenly crossed the upload/download signals in my head), probably fine to leave the comment as-is too. :+1:

vvoland commented on daemon/containerd/image_push.go:
```diff
@@ -2,15 +2,277 @@ package containerd
+// Otherwise index is pushed as is, which will probably fail if you don't have all
+// content referenced by the index. Cross-repo mounts will be attempted for non-existing blobs.
+func (i *ImageService) pushImage(ctx context.Context, targetRef reference.Named, optPlatform *ocispec.Platform, metaHeaders map[string][]string, authConfig *registry.AuthConfig, outStream io.Writer) error {
+	if _, ok := targetRef.(reference.Tagged); !ok {
+		return errdefs.NotImplemented(errors.New("push all tags is not implemented"))
+	}
```
I'm kinda wondering if push "all" tags is still useful with the multi-platform images. I can imagine it being useful when you had to do separate tags for multi-platform images (eg `alpine:amd64`, `alphine:arm64` etc), but with multi-platform images there might be less use cases for it?
Obviously, I might be missing some more context behind it, so happy to hear what others think 😅

vvoland commented:
Removed `platform` parameter and related code. Also added the code to append distribution source layers to the pushed content. Can remove it if it's too much though.

thaJeztah commented on daemon/containerd/image_push.go:
```diff
@@ -2,14 +2,256 @@ package containerd
+	return err
+}
+
+// prepareMissing will walk the target descriptor recursively and returns up
```
Looks like the docs doesn't match the function name

vvoland replied:
Thanks! Fixed.

thaJeztah commented on daemon/containerd/image_push.go:
```diff
@@ -2,14 +2,256 @@ package containerd
+	return reference.WithDigest(source.registryRef, dgst)
+}
+
+// shouldDownload returns if the given descriptor can be cross-repo mounted
```
Same here

vvoland replied:
Thanks! Fixed.

thaJeztah reviewed (approved):
"LGTM" from a UX perspective (as discussed) I'm good with bringing this one in, but would appreciate a few eyes on the code

tianon commented on daemon/containerd/progress.go:
```diff
@@ -158,3 +159,76 @@ func (p pullProgress) UpdateProgress(ctx context.Context, ongoing *jobs, out pro
+
+		if status.Committed && status.Offset >= status.Total {
+			if p.isMountable(j.Digest) {
+				progress.Update(out, id, "Mounted")
```
I think the current output from `docker push` on a cross-blob mount specifies _where_ it was mounted from; is it possible to surface that data somehow here as well? (is that already in here and I missed it? :see_no_evil:)

tianon replied:
Aha, found the ref: https://github.com/moby/moby/blob/0656059ae7ab1d0b14afd3bd5f4533f2f2083ff0/distribution/push_v2.go#L379

tianon reviewed (approved):
One minor comment, but probably fine for a follow-up :+1:

thaJeztah commented on daemon/containerd/image_push.go:
```diff
@@ -2,15 +2,255 @@ package containerd
+	// Check if this blob has a distributionSource label
+	// if yes, read it as source
+	for k, v := range labels {
+		registry := strings.TrimPrefix(k, labelDistributionSource)
+		if registry != k {
+			ref, err := reference.ParseNamed(registry + "/" + v)
```
nit: Looks like `registry` collides with the `"github.com/docker/docker/api/types/registry"` import. perhaps `reg` is an option (perhaps also inline it, as it's only used inside the `if`);

```suggestion
		if reg := strings.TrimPrefix(k, labelDistributionSource); reg != k {
			ref, err := reference.ParseNamed(reg + "/" + v)
```

vvoland replied:
Done in: https://github.com/moby/moby/pull/45243

thaJeztah commented on daemon/containerd/image_push.go:
```diff
@@ -2,15 +2,255 @@ package containerd
+		return false
+	}
+
+	registry := reference.Domain(targetRef)
```
same here (`registry` colliding with import)

vvoland replied:
Done in: https://github.com/moby/moby/pull/45243

thaJeztah reviewed (commented):
Looked a bit at the code; overall looks good, but left some nits, and a question about concurrency

thaJeztah commented:
(those should be fine for a follow-up though 😂 because it's already in 😂)

vvoland commented on daemon/containerd/image_push.go:
```diff
@@ -2,15 +2,255 @@ package containerd
+	if err != nil {
+		return err
+	}
+	defer release(leasedCtx)
```
Technically, if it returned an error then the lease couldn't be deleted, so that might be something worthy of logging. However, looking at how containerd itself uses this, it's "common" (I haven't really seen any code that does it differently) to just call it and ignore the error.
Looking at the code, the only case where this could happen if the lease deletion fails. This can happen if the lease no longer exists (which is already the state we want) or the c8d connection is broken, which will pop out in some other place anyway.

thaJeztah replied:
Gotcha; yeah, TBH, containerd ignoring things could _sometimes_ be a guideline, but we should definitely check ourselves what makes sense (mistakes are made by everyone, so it's always possible containerd is on the wrong on these, in which case we could contribute there as well).

- "not found" errors; yup definitely we should ignore
- more curious; could "something else using the files / mounts / whatever" cause things to not be cleaned up? (in which case we may want to have a log to backtrace we potentially leaked things somewhere); of course that is if that wouldn't trigger "false positives" (race conditions, things being "eventually" cleaned up etc).

vvoland replied:
This only deletes a lease - which is created only in this call-site and deleting it shouldn't fail unless the lease was already deleted, containerd connection is broken or containerd state is corrupted. In which case we will have loud errors in other places. Even if we fail to delete the lease and containerd is fully operational with the lease state still being present, the lease will expire after 24h.
It doesn't harm to log it though.. Except it's just bloating a simple one line defer to a deferred anonymous function call with a conditional log invocation 😅  But maybe it can save us from some trouble if the Lease manager has some major future changes from the current state, so I'm not that much against logging it.

thaJeztah replied:
Thanks! Yes I didn't know if actual _mounts_ and/or _files_ were involved here. I'm mostly "fine" either way I guess

vvoland commented on daemon/containerd/image_push.go:
```diff
@@ -2,15 +2,255 @@ package containerd
+	store := i.client.ContentStore()
+
+	resolver, tracker := i.newResolverFromAuthConfig(authConfig)
+	pushProgress := pushProgress{Tracker: tracker}
```
Done in https://github.com/moby/moby/pull/45243

vvoland commented on daemon/containerd/image_push.go:
```diff
@@ -2,15 +2,255 @@ package containerd
+			if err != nil {
+				return nil, err
+			}
+			for _, c := range children {
+				jobs.Add(c)
+			}
```
Done in https://github.com/moby/moby/pull/45243

vvoland commented on daemon/containerd/image_push.go:
```diff
@@ -2,15 +2,255 @@ package containerd
+
+			for _, source := range sources {
+				if canBeMounted(desc.MediaType, targetRef, i.registryService.IsInsecureRegistry, source) {
+					mountableBlobs[desc.Digest] = source
```
Yes, good catch! This is explicitly called concurrently due to `Dispatch` instead of `Walk` so there's a data race here.
